infra/gravity/terraform.go

```diff
@@ -0,0 +1,131 @@
+package gravity
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/gravitational/robotest/infra"
+	"github.com/gravitational/robotest/infra/terraform"
+	"github.com/gravitational/robotest/lib/utils"
+
+	"github.com/gravitational/trace"
+)
+
+// DestroyFn function which will destroy previously created remote resources
+type DestroyFn func(context.Context, *testing.T) error
+
+var destroyOnSuccess = flag.Bool("destroy-on-success", true, "remove resources after test success")
+var destroyOnFailure = flag.Bool("destroy-on-failure", false, "remove resources after test failure")
+
+var resourceListFile = flag.String("resourcegroupfile", "", "file with list of resources created")
+
+var testStatus = map[bool]string{true: "failed", false: "ok"}
+
+// wrapDestroyFn implements a global conditional logic
+func wrapDestroyFn(tag string, destroy func(context.Context) error) DestroyFn {
+	return func(ctx context.Context, t *testing.T) error {
+		log := utils.Logf(t, tag)
+
+		if (*destroyOnSuccess == false) ||
+			(t.Failed() && *destroyOnFailure == false) {
+			log("Not removing terraform %s")
+			return nil
+		}
+
+		log("destroying Terraform resources %s, test %s", tag, testStatus[t.Failed()])
+
+		err := destroy(ctx)
+		if err != nil {
+			log("error destroying terraform resources: %v", err)
+		} else {
+			resourceDestroyed(tag)
+		}
+
+		return trace.Wrap(err)
+	}
+}
+
+var resourceAllocations = struct {
+	sync.Mutex
+	tags map[string]bool
+}{tags: map[string]bool{}}
+
+// resourceAllocated adds resource allocated into local index file for shell-based cleanup
+// as test might crash and leak resources on the cloud
+func resourceAllocated(tag string) error {
+	resourceAllocations.Lock()
+	defer resourceAllocations.Unlock()
+
+	if _, there := resourceAllocations.tags[tag]; there {
+		return trace.Errorf("resource tag not unique : %s", tag)
+	}
+
+	resourceAllocations.tags[tag] = true
+	return saveResourceAllocations()
+}
+
+func resourceDestroyed(tag string) error {
+	resourceAllocations.Lock()
+	defer resourceAllocations.Unlock()
+
+	delete(resourceAllocations.tags, tag)
+	return saveResourceAllocations()
+}
+
+func saveResourceAllocations() error {
+	if *resourceListFile == "" {
+		return nil
+	}
+
+	file, err := os.OpenFile(*resourceListFile, os.O_RDWR|os.O_CREATE, 0755)
```

Comment:
https://github.com/gravitational/robotest/blob/master/lib/constants/constants.go#L17 or create a new one for files